File: wrappers/dotnet/indy-sdk-dotnet/Wrapper/CustomWalletBase.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace Indy.Sdk.Dotnet.Wrapper
{
    /// <summary>
    /// Base type for custom wallets.
    /// </summary>
    public abstract class CustomWalletBase
    {
        /// <summary>
        /// Pointers to values that have been allocated to unmanaged memory by the wallet.
        /// </summary>
        private List<IntPtr> _valuePointers = new List<IntPtr>();

        /// <summary>
        /// Gets a list of active unmanaged pointers for the wallet.
        /// </summary>
        internal List<IntPtr> ValuePointers
        {
            get { return _valuePointers; }
        }

        /// <summary>
        ///  Allows an implementer to set a value in the wallet.
        /// </summary>
        /// <param name="key">The key of the value to set.</param>
        /// <param name="value">The value to set.</param>
        public abstract ErrorCode Set(string key, string value);

        /// <summary>
        /// Allows an implementer to get a value from the wallet.
        /// </summary>
        /// <param name="key">The key of the value to get.</param>
        /// <param name="value">The value obtained from the wallet.</param>
        public abstract ErrorCode Get(string key, out string value);

        /// <summary>
        /// Allows an implementer to get a value from the wallet if it has not expired.
        /// </summary>
        /// <param name="key">The key of the value to get.</param>
        /// <param name="value">The value obtained from the wallet.</param>
        public abstract ErrorCode GetNotExpired(string key, out string value);

        /// <summary>
        /// Allows an implementer to get a list of values from the wallet that match a key prefix.
        /// </summary>
        /// <param name="keyPrefix">The key prefix for the values requested.</param>
        /// <param name="valuesJson">The JSON string containing the values associated with the key prefix.</param>
        public abstract ErrorCode List(string keyPrefix, out string valuesJson);
        
        /// <summary>
        /// Disposes a wallet instance and ensures any remaining unmanaged pointers are freed.
        /// </summary>
        public void Dispose()
        {
            //Free any outstanding handles.
            for (int i = _valuePointers.Count - 1; i >= 0; i--)
            {
                Marshal.FreeHGlobal(_valuePointers[i]);
                _valuePointers.RemoveAt(i);
            }
        }
    }
}

File: wrappers/java/src/main/java/org/hyperledger/indy/sdk/wallet/CustomWalletBase.java
package org.hyperledger.indy.sdk.wallet;

import java.util.ArrayList;
import java.util.List;

import org.hyperledger.indy.sdk.ErrorCode;
import org.hyperledger.indy.sdk.wallet.WalletType.StringByReference;

import com.sun.jna.Native;
import com.sun.jna.Pointer;

/**
 * All custom wallets must inherit from this base class.
 */
public abstract class CustomWalletBase implements AutoCloseable {

	/// <summary>
    /// Pointers to values that have been allocated to unmanaged memory by the wallet.
    /// </summary>
    private List<Pointer> valuePointers = new ArrayList<Pointer>();
    
    public List<Pointer> getValuePointers(){
    	return valuePointers;
    }

	/**
	 * Sets a value on a wallet instance.
	 * 
	 * @param handle The handle of the wallet to set the value on.
	 * @param key The key to set the value for.
	 * @param value The value to set.
	 * @return An ErrorCode indicating the outcome.
	 */
	public abstract ErrorCode set(String key, String value);
	
	/**
	 * Gets a value from a wallet instance.
	 * 
	 * @param handle The handle of the wallet to set the value on.
	 * @param key The key of value to get.
	 * @param resultString A result object to containthe value set by implementers.
	 * @return An ErrorCode indicating the outcome.
	 */
	public abstract ErrorCode get(String key, StringByReference resultString);
	
	/**
	 * Gets an unexpired value from a wallet instance.
	 * 
	 * @param handle The handle of the wallet to set the value on.
	 * @param key The key of value to get.
	 * @param resultString A result object to containthe value set by implementers.
	 * @return An ErrorCode indicating the outcome.
	 */
	public abstract ErrorCode getNotExpired(String key, StringByReference resultString);
	
	/**
	 * Gets a list of values optionally filtered by key.
	 * 
	 * @param handle The handle of the wallet to set the value on.
	 * @param keyPrefx The prefix of the keys to filter on.  If null no filter will be applied.
	 * @param resultString A result object to containthe value set by implementers.
	 * @return An ErrorCode indicating the outcome.
	 */
	public abstract ErrorCode list(String keyPrefix, StringByReference resultString);
    
	
	/**
	 * Closes the wallet.
	 */
	@Override
	public void close() throws Exception {
		//Free any outstanding handles.
        for (int i = valuePointers.size() - 1; i >= 0; i--)
        {
        	Pointer valuePointer = valuePointers.get(i);
        	Native.free(Pointer.nativeValue(valuePointer));
            valuePointers.remove(i);
        }
	}

}
